core/packages/contracts/src/BeefyClient.sol

```diff
@@ -301,6 +336,10 @@ contract BeefyClient is Ownable {
 
         Ticket storage ticket = tickets[ticketID];
 
+        if (ticket.validatorSetLen != currentValidatorSet.length) {
+            revert InvalidTicket();
+        }
+
         if (commitment.validatorSetID != currentValidatorSet.id) {
```

Comment:
Hmm, doesn't this check here prevent relayers from submitting commitments signed by an invalid validator set?

Validator set ids are monotonically increasing. So `nextValidatorSet.id + 1 == currentValidatorSet.id`.

Reply:
It does, but the attack we're defending against here already uses the correct validator set in the `submitFinal` call:

- The attacker uses the `submitInitial` call for the smaller validator set, so they only need to supply a bitfield with 2/3 of the smaller set's size in set bits.
(They also need to compromise one validator in the smaller set, and ensure that they've compromised the validator with the corresponding index in the larger validator set. The compromised validator also needs to sign the commitment for another era, so we should be able to slash them afterwards.)
- When the attacker gets to `submitFinal`, they can use the larger (& valid/correct) validator set. The id check here passes, because the set ids match. But now we're sampling from a bitfield that only has set bits equal to 2/3 of the smaller set's size, not of the larger as it should.

Come to think of it, if we did this id check in `submitInitial*` we'd also defend against this attack. The commitment hash is used to look up the ticket, so the commitment needs to be the same across submitInitial & submitFinal. But the commitment has to have a single validator set id - either current or next - so mixing handover & non-handover calls for the same commitment would fail one of the commitment id checks.

Reply:
> The attacker uses the submitInitial call for the smaller validator set

AFAIK active validators are elected from polkadot staking with a fixed size(297)  now and will increased to 1000 in long run. So is that possible we add another check when update validator set here.

https://github.com/Snowfork/snowbridge/blob/d26678debcac6048c2c7dd2db54c06687575661e/core/packages/contracts/src/BeefyClient.sol#L355

Something like `assert(nextValidatorSet.length>=currentValidatorSet.length)`?

Reply:
> It does, but the attack we're defending against here already uses the correct validator set in the submitFinal call:

ah I see, I understand better now.

> Come to think of it, if we did this id check in submitInitial* we'd also defend against this attack. The commitment hash is used to look up the ticket, so the commitment needs to be the same across submitInitial & submitFinal. But the commitment has to have a single validator set id - either current or next - so mixing handover & non-handover calls for the same commitment would fail one of the commitment id checks.

I agree, I think this this a more comprehensive check. Let's go this route rather.

> AFAIK active validators are elected from polkadot [staking](https://polkadot.js.org/apps/?rpc=wss%3A%2F%2Fpolkadot-rpc-tn.dwellir.com#/staking) with a fixed size(297) now and will increased to 1000 in long run. So is that possible we add another check when update validator set [here](https://github.com/Snowfork/snowbridge/blob/d26678debcac6048c2c7dd2db54c06687575661e/core/packages/contracts/src/BeefyClient.sol#L355) something like assert(nextValidatorSet.length>=currentValidatorSet.length)?

This would not be needed if we check the validator set id in submitInitial, as per David's suggestion